sql/src/main/java/io/crate/execution/engine/join/RamBlockSizeCalculator.java

```diff
@@ -51,15 +51,19 @@ public int getAsInt() {
         if (statisticsUnavailable(circuitBreaker, estimatedRowSizeForLeft, numberOfRowsForLeft)) {
             return defaultBlockSize;
         }
+        long availableMemory = circuitBreaker.getLimit() - circuitBreaker.getUsed();
+        long numRowsFittingIntoAvailableMemory = availableMemory / estimatedRowSizeForLeft;
 
-        int blockSize = (int) Math.min(Integer.MAX_VALUE, (circuitBreaker.getLimit() - circuitBreaker.getUsed()) / estimatedRowSizeForLeft);
-        blockSize = (int) Math.min(numberOfRowsForLeft, blockSize);
-
-        // for distributed hash joins, we must ensure that each parallel executed join is switching to the same relation
-        // eventually and does not load the next batch while another is already switching. this would result in a
-        // dead lock caused by the constraint that all receivers must response to the collect nodes before a next batch
-        // is sent.
-        blockSize = Math.min(defaultBlockSize, blockSize);
+        // Restrict the number of rows per block by whatever is lowest:
```

Comment:
this is very good and needed here

Reply:
I'll push another fixup to eliminate the `Integer.MAX_VALUE` thing - the `defaultBlockSize` is already an `int` so it can't be larger that that anyways.